File: api/src/routes/documents.ts
import { Hono } from 'hono';
import { Env } from '../types';
import { requireAuth } from '../auth';
import { analyzeDocumentContent } from '../services/ai';

const router = new Hono<{ Bindings: Env }>();

router.post('/upload', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const formData = await c.req.formData();
  const file = formData.get('file') as File;
  const title = (formData.get('title') as string) || file.name;

  if (!file) return c.json({ detail: 'No file provided' }, 400);

  const key = `${user.id}/${Date.now()}-${file.name}`;
  await c.env.DOCUMENTS.put(key, file);

  const { meta } = await c.env.DB.prepare(
    'INSERT INTO documents (user_id, title, file_path, file_type, doc_type, status) VALUES (?, ?, ?, ?, ?, ?)'
  ).bind(user.id, title, key, file.name.split('.').pop(), 'other', 'completed').run();

  const doc = await c.env.DB.prepare('SELECT id, title, status FROM documents WHERE id = ?').bind(meta.last_row_id).first();
  return c.json({ message: 'Document uploaded', document: doc });
});

router.post('/analyze/:id', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const docId = c.req.param('id');
  const { content } = await c.req.json();

  const doc = await c.env.DB.prepare('SELECT * FROM documents WHERE id = ? AND user_id = ?').bind(docId, user.id).first();
  if (!doc) return c.json({ detail: 'Document not found' }, 404);

  const analysis = analyzeDocumentContent(content || '');
  await c.env.DB.prepare('UPDATE documents SET ai_analysis = ?, risk_score = ?, status = ? WHERE id = ?')
    .bind(JSON.stringify(analysis), analysis.risk_score, 'completed', docId).run();

  return c.json(analysis);
});

router.get('/', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const docs = await c.env.DB.prepare('SELECT id, title, doc_type, status, risk_score, created_at FROM documents WHERE user_id = ? ORDER BY created_at DESC').bind(user.id).all();
  return c.json(docs.results);
});

router.get('/:id', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const doc = await c.env.DB.prepare('SELECT * FROM documents WHERE id = ? AND user_id = ?').bind(c.req.param('id'), user.id).first();
  if (!doc) return c.json({ detail: 'Document not found' }, 404);
  return c.json(doc);
});

router.delete('/:id', async (c) => {
  const user = await requireAuth(c.req.header('authorization'), c.env, c.env.DB);
  const doc = await c.env.DB.prepare('SELECT * FROM documents WHERE id = ? AND user_id = ?').bind(c.req.param('id'), user.id).first();
  if (!doc) return c.json({ detail: 'Document not found' }, 404);
  await c.env.DB.prepare('DELETE FROM documents WHERE id = ?').bind(c.req.param('id')).run();
  return c.json({ message: 'Document deleted' });
});

export default router;

File: backend/app/routes/documents.py
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File, Form
from sqlalchemy.orm import Session
from app.database import get_db
from app.models.document import Document, DocumentType, DocumentStatus
from app.models.user import User
from app.utils.auth import get_current_user
from app.utils.documents import save_upload_file, analyze_document_content

router = APIRouter(prefix="/api/documents", tags=["Documents"])

@router.post("/upload")
async def upload_document(
    file: UploadFile = File(...),
    title: str = Form(None),
    doc_type: str = Form("other"),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    try:
        file_path = await save_upload_file(file)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    doc = Document(
        user_id=current_user.id,
        title=title or file.filename,
        file_path=file_path,
        file_type=file.filename.split(".")[-1],
        doc_type=doc_type,
        status=DocumentStatus.COMPLETED,
    )
    db.add(doc)
    db.commit()
    db.refresh(doc)
    return {"message": "Document uploaded", "document": {"id": doc.id, "title": doc.title, "status": doc.status.value}}

@router.post("/analyze/{doc_id}")
def analyze_document(
    doc_id: int,
    content: str = Form(""),
    current_user: User = Depends(get_current_user),
    db: Session = Depends(get_db),
):
    doc = db.query(Document).filter(Document.id == doc_id, Document.user_id == current_user.id).first()
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    analysis = analyze_document_content(content)
    doc.ai_analysis = str(analysis)
    doc.risk_score = analysis["risk_score"]
    doc.status = DocumentStatus.COMPLETED
    db.commit()
    return analysis

@router.get("/")
def list_documents(current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    docs = db.query(Document).filter(Document.user_id == current_user.id).all()
    return [
        {
            "id": d.id,
            "title": d.title,
            "doc_type": d.doc_type.value if d.doc_type else None,
            "status": d.status.value if d.status else None,
            "risk_score": d.risk_score,
            "created_at": str(d.created_at),
        }
        for d in docs
    ]

@router.get("/{doc_id}")
def get_document(doc_id: int, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    doc = db.query(Document).filter(Document.id == doc_id, Document.user_id == current_user.id).first()
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    return {
        "id": doc.id,
        "title": doc.title,
        "file_type": doc.file_type,
        "doc_type": doc.doc_type.value if doc.doc_type else None,
        "status": doc.status.value if doc.status else None,
        "risk_score": doc.risk_score,
        "ai_analysis": doc.ai_analysis,
        "created_at": str(doc.created_at),
    }

@router.delete("/{doc_id}")
def delete_document(doc_id: int, current_user: User = Depends(get_current_user), db: Session = Depends(get_db)):
    doc = db.query(Document).filter(Document.id == doc_id, Document.user_id == current_user.id).first()
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    db.delete(doc)
    db.commit()
    return {"message": "Document deleted"}
